File: oop.js
/* OOP (Object Oriented Programming) */

class Figura {
    constructor(){
        console.log("Instanciando la clase Figura")
        this.area = 0;
    }

    calcularArea(){}

    getArea(){
        return this.area;
    }
}
/* 
let fig = new Figura();
fig.calcularArea()
fig.getArea() 
*/

class Cuadrado extends Figura {
    constructor(){
        super();
        this.lado = 0;
    }

    calcularArea(lado){
        this.area = lado*lado;
    }
}

let cuadrado = new Cuadrado();
cuadrado.calcularArea(10)
console.log(cuadrado.getArea())

class Triangulo extends Figura {
    constructor(base, altura){
        super();
        this.base = base;
        this.altura = altura;
    }

    calcularArea(){
        this.area = (this.base * this.altura) / 2;
    }
}

let triangulo = new Triangulo(10, 15);
triangulo.calcularArea();
console.log(triangulo.getArea())


class Libro {
    constructor(titulo, autor){
        this.titulo = titulo;
        this.autor = autor;
    }

    mostrarInformacion(){
        return `Titulo: ${this.titulo}\nAutor: ${this.autor}`;
    }
}

let lib1 = new Libro("El Gran Gatsby", "Fitzgerald");
let lib2 = new Libro("100 Años de Soledad", "Gabriel Garcia Marquez");
let lib3 = new Libro("Harry Potter", "J.K Rowling");
//console.log(lib1.mostrarInformacion())


class Biblioteca {
    constructor(nombre){
        this.nombre = nombre;
        this.libros = [];
    }

    agregarLibro(libro){
        this.libros.push(libro)
    }

    mostrarCatalogoLibros(){
        if(this.libros.length > 0){
            this.libros.forEach((libro) => {
                console.log(libro.mostrarInformacion())
            })
        } else {
            console.log("Sin disponibilidad de Libros")
        }
    }
}

let biblioteca = new Biblioteca("Biblioteca Nacional de Chile");
biblioteca.mostrarCatalogoLibros()
biblioteca.agregarLibro(lib1)
biblioteca.agregarLibro(lib2)
biblioteca.agregarLibro(lib3)
biblioteca.mostrarCatalogoLibros()

File: oop.py
""" OOP (Object Oriented Programming) """

class Figura:
    def __init__(self):
        print("Instanciando la clase Figura")
        self.area = 0
        
    def calcular_area(self):
        pass
    
    def getArea(self):
        return self.area
        
"""         
fig = Figura()
fig.calcular_area()
fig.getArea()
"""

class Cuadrado(Figura):
    def __init__(self):
        super().__init__()
        self.lado = 0
        
    def calcular_area(self, lado):
        self.area = lado * lado
        

cuadrado = Cuadrado()
cuadrado.calcular_area(10)
print(cuadrado.getArea())


class Triangulo(Figura):
    def __init__(self, base, altura):
        super().__init__()
        self.base = base
        self.altura = altura
        
    def calcular_area(self):
        self.area = (self.base * self.altura) / 2
        
        
triangulo = Triangulo(20, 14)
triangulo.calcular_area()
print(triangulo.getArea())


class Libro:
    def __init__(self, titulo, autor):
        self.titulo = titulo
        self.autor = autor
        
    def mostrar_informacion(self):
        return f"Titulo: {self.titulo}\nAutor: {self.autor}"
    
lib1 = Libro("El Gran Gatsby", "Fitzgerald");
lib2 = Libro("100 Años de Soledad", "Gabriel Garcia Marquez");
lib3 = Libro("Harry Potter", "J.K Rowling");


class Biblioteca:
    def __init__(self, nombre):
        self.nombre = nombre
        self.libros = []
        
    def agregar_libro(self, libro):
        self.libros.append(libro)
        
    def mostrar_catalogo_libros(self):
        if len(self.libros) > 0:
            for libro in self.libros:
                print(libro.mostrar_informacion())
        else:
            print("Sin disponibilidad de Libros")
            
            
biblioteca = Biblioteca("Biblioteca Nacional de Chile")
biblioteca.mostrar_catalogo_libros()
biblioteca.agregar_libro(lib1)
biblioteca.agregar_libro(lib2)
biblioteca.agregar_libro(lib3)
biblioteca.mostrar_catalogo_libros()